Fix MySQL bugs related to only_full_group_by

Alkarex opened the pull request:
Fixes #3322

Alkarex commented:
I can see in the code that other places have wrong GROUP BYs, which would need the same kind of fix, but the one fixed here was crashing my cron :-P

annando commented:
Thank you very much for pointing at this MySQL problem that I wasn't aware of!

But I have the feeling that adding these fields to the "group by" could change the content of the result set, couldn't it? I would like to make some tests, but currently I'm out of time.

Alkarex commented:
It would be good to test, of course, but that should be backward compatible.
See the explanation https://dev.mysql.com/doc/refman/8.0/en/group-by-handling.html
In a nutshell, to work everywhere, a GROUP BY must contain all the fields that are used in the SELECT and/or ORDER BY clauses. When not written explicitly,  I believe this is done implicitly by SQL engines anyway.

Examples from my instance:

```sh
mysql> SELECT `term`.`oid`, `term`.`url`, `term`.`uid` FROM `term`
    -> STRAIGHT_JOIN `thread` ON `thread`.`iid` = `term`.`oid`
    -> AND `thread`.`uid` = `term`.`uid`
    -> WHERE `term`.`type` = 7 AND `term`.`term` > '2017-04-01 21:40:03'
    -> AND `thread`.`mention`
    -> GROUP BY `term`.`url`, `term`.`uid`, `term`.`oid`, `term`.`term`
    -> ORDER BY `term`.`term` DESC;
+-----+--------------------------------------------------+-----+
| oid | url                                              | uid |
+-----+--------------------------------------------------+-----+
| 388 | https://ostatus.shnoulle.net/conversation/578534 |   1 |
|   1 | https://s.a80.fr/conversation/1                  |   1 |
| 276 | https://herds.eu/conversation/1913222            |   1 |
+-----+--------------------------------------------------+-----+
3 rows in set (0,02 sec)

mysql> SELECT `oid`, `url`, `uid` FROM `term` WHERE `type` = 7
    -> AND `term` > '2017-04-10 23:00:09' GROUP BY `url`, `uid`, `oid`, `term`
    -> ORDER BY `term` DESC;
+-----+-------------------------------------------------------+-----+
| oid | url                                                   | uid |
+-----+-------------------------------------------------------+-----+
| 569 | https://ostatus.shnoulle.net/conversation/578870      |   1 |
| 567 | https://ostatus.shnoulle.net/conversation/578868      |   1 |
| 563 | https://ostatus.shnoulle.net/conversation/578810      |   1 |
| 565 | https://ostatus.shnoulle.net/conversation/578737      |   1 |
| 559 | https://ostatus.shnoulle.net/conversation/578737      |   1 |
| 551 | https://ostatus.shnoulle.net/conversation/578772      |   1 |
| 541 | https://ostatus.shnoulle.net/conversation/578810      |   1 |
| 537 | https://ostatus.shnoulle.net/conversation/578806      |   1 |
| 543 | https://social.umeahackerspace.se/conversation/662677 |   1 |
| 535 | https://ostatus.shnoulle.net/conversation/578795      |   1 |
| 533 | https://ostatus.shnoulle.net/conversation/578503      |   1 |
| 523 | https://ostatus.shnoulle.net/conversation/578731      |   1 |
| 517 | https://ostatus.shnoulle.net/conversation/578731      |   1 |
| 511 | https://ostatus.shnoulle.net/conversation/578737      |   1 |
| 515 | https://ostatus.shnoulle.net/conversation/578731      |   1 |
| 505 | https://ostatus.shnoulle.net/conversation/1           |   1 |
| 499 | https://ostatus.shnoulle.net/conversation/578731      |   1 |
| 497 | https://ostatus.shnoulle.net/conversation/578725      |   1 |
| 501 | https://ostatus.shnoulle.net/conversation/578720      |   1 |
| 493 | https://ostatus.shnoulle.net/conversation/578716      |   1 |
| 495 | https://ostatus.shnoulle.net/conversation/578711      |   1 |
| 545 | https://social.umeahackerspace.se/conversation/662612 |   1 |
| 489 | https://social.umeahackerspace.se/conversation/662586 |   1 |
| 485 | https://quitter.is/conversation/1997362               |   1 |
| 467 | https://ostatus.shnoulle.net/conversation/578669      |   1 |
| 487 | https://social.umeahackerspace.se/conversation/662586 |   1 |
| 477 | https://social.umeahackerspace.se/conversation/662033 |   1 |
| 475 | https://social.umeahackerspace.se/conversation/662033 |   1 |
| 483 | https://quitter.is/conversation/1997362               |   1 |
| 481 | https://quitter.is/conversation/1997362               |   1 |
| 449 | https://ostatus.shnoulle.net/conversation/578590      |   1 |
| 445 | https://social.umeahackerspace.se/conversation/662586 |   1 |
| 441 | https://herds.eu/conversation/1921352                 |   1 |
| 435 | https://ostatus.shnoulle.net/conversation/578318      |   1 |
| 439 | https://herds.eu/conversation/1921352                 |   1 |
| 447 | https://ostatus.shnoulle.net/conversation/578590      |   1 |
| 473 | https://social.umeahackerspace.se/conversation/662033 |   1 |
| 471 | https://social.umeahackerspace.se/conversation/662033 |   1 |
| 396 | https://ostatus.shnoulle.net/conversation/578542      |   1 |
| 388 | https://ostatus.shnoulle.net/conversation/578534      |   1 |
| 380 | https://ostatus.shnoulle.net/conversation/578318      |   1 |
| 376 | https://ostatus.shnoulle.net/conversation/577490      |   1 |
| 372 | https://ostatus.shnoulle.net/conversation/578227      |   1 |
| 370 | https://ostatus.shnoulle.net/conversation/578227      |   1 |
| 362 | https://ostatus.shnoulle.net/conversation/578473      |   1 |
| 352 | https://ostatus.shnoulle.net/conversation/578513      |   1 |
| 348 | https://ostatus.shnoulle.net/conversation/574610      |   1 |
| 274 | https://ostatus.shnoulle.net/conversation/576472      |   1 |
| 346 | https://ostatus.shnoulle.net/conversation/574610      |   1 |
| 531 | https://ostatus.shnoulle.net/conversation/578503      |   1 |
| 350 | https://ostatus.shnoulle.net/conversation/578513      |   1 |
| 354 | https://ostatus.shnoulle.net/conversation/578495      |   1 |
| 412 | http://quitter.se/conversation/7098248                |   1 |
| 360 | https://ostatus.shnoulle.net/conversation/578473      |   1 |
| 270 | https://ostatus.shnoulle.net/conversation/576472      |   1 |
| 404 | https://ostatus.shnoulle.net/conversation/578011      |   1 |
| 248 | https://ostatus.shnoulle.net/conversation/578011      |   1 |
| 246 | https://social.umeahackerspace.se/conversation/660325 |   1 |
| 204 | https://herds.eu/conversation/1921352                 |   1 |
| 122 | https://social.umeahackerspace.se/conversation/659617 |   1 |
| 244 | https://s.a80.fr/conversation/80                      |   1 |
| 368 | https://ostatus.shnoulle.net/conversation/578227      |   1 |
| 242 | https://herds.eu/conversation/1921352                 |   1 |
| 240 | https://herds.eu/conversation/1921352                 |   1 |
| 238 | https://herds.eu/conversation/1921352                 |   1 |
| 236 | https://herds.eu/conversation/1921352                 |   1 |
| 234 | https://herds.eu/conversation/1921352                 |   1 |
| 231 | https://herds.eu/conversation/1921352                 |   1 |
| 423 | https://ostatus.shnoulle.net/conversation/577334      |   1 |
| 410 | https://ostatus.shnoulle.net/conversation/578068      |   1 |
| 227 | https://herds.eu/conversation/1921352                 |   1 |
| 222 | https://herds.eu/conversation/1921352                 |   1 |
| 252 | https://ostatus.shnoulle.net/conversation/578068      |   1 |
| 402 | https://ostatus.shnoulle.net/conversation/578011      |   1 |
| 400 | https://ostatus.shnoulle.net/conversation/578011      |   1 |
| 398 | https://ostatus.shnoulle.net/conversation/578011      |   1 |
| 220 | https://herds.eu/conversation/1921352                 |   1 |
| 218 | https://herds.eu/conversation/1921352                 |   1 |
| 216 | https://herds.eu/conversation/1921352                 |   1 |
| 214 | https://herds.eu/conversation/1921352                 |   1 |
| 205 | http://quitter.se/conversation/7098248                |   1 |
| 394 | https://ostatus.shnoulle.net/conversation/577776      |   1 |
| 202 | http://quitter.se/conversation/7098248                |   1 |
| 392 | https://ostatus.shnoulle.net/conversation/577776      |   1 |
|  68 | https://ostatus.shnoulle.net/conversation/577678      |   1 |
| 200 | http://quitter.se/conversation/7098248                |   1 |
| 198 | http://quitter.se/conversation/7098248                |   1 |
| 196 | http://quitter.se/conversation/7098248                |   1 |
| 194 | http://quitter.se/conversation/7098248                |   1 |
| 192 | http://quitter.se/conversation/7098248                |   1 |
| 269 | https://ostatus.shnoulle.net/conversation/577490      |   1 |
|  29 | https://ostatus.shnoulle.net/conversation/577490      |   1 |
| 190 | http://quitter.se/conversation/7098248                |   1 |
|  17 | https://qvitter.krashboyz.org/conversation/41767      |   1 |
| 420 | https://qvitter.krashboyz.org/conversation/41767      |   1 |
| 186 | http://quitter.se/conversation/7098248                |   1 |
| 184 | http://quitter.se/conversation/7098248                |   1 |
| 182 | http://quitter.se/conversation/7098248                |   1 |
| 180 | http://quitter.se/conversation/7098248                |   1 |
| 178 | http://quitter.se/conversation/7098248                |   1 |
| 176 | http://quitter.se/conversation/7098248                |   1 |
| 174 | http://quitter.se/conversation/7098248                |   1 |
|  15 | https://qvitter.krashboyz.org/conversation/41767      |   1 |
| 419 | https://qvitter.krashboyz.org/conversation/41767      |   1 |
| 172 | http://quitter.se/conversation/7098248                |   1 |
| 170 | http://quitter.se/conversation/7098248                |   1 |
| 168 | http://quitter.se/conversation/7098248                |   1 |
| 166 | http://quitter.se/conversation/7098248                |   1 |
| 164 | http://quitter.se/conversation/7098248                |   1 |
| 268 | https://ostatus.shnoulle.net/conversation/577490      |   1 |
|  27 | https://ostatus.shnoulle.net/conversation/577490      |   1 |
| 406 | https://social.umeahackerspace.se/conversation/660325 |   1 |
| 230 | https://social.umeahackerspace.se/conversation/660093 |   1 |
| 162 | http://quitter.se/conversation/7098248                |   1 |
| 160 | http://quitter.se/conversation/7098248                |   1 |
| 421 | https://social.umeahackerspace.se/conversation/660093 |   1 |
| 158 | http://quitter.se/conversation/7098248                |   1 |
| 156 | http://quitter.se/conversation/7098248                |   1 |
| 154 | http://quitter.se/conversation/7098248                |   1 |
| 152 | http://quitter.se/conversation/7098248                |   1 |
| 267 | https://ostatus.shnoulle.net/conversation/577490      |   1 |
|  25 | https://ostatus.shnoulle.net/conversation/577490      |   1 |
| 266 | https://ostatus.shnoulle.net/conversation/577490      |   1 |
|  23 | https://ostatus.shnoulle.net/conversation/577490      |   1 |
|  13 | https://qvitter.krashboyz.org/conversation/41767      |   1 |
| 418 | https://qvitter.krashboyz.org/conversation/41767      |   1 |
|  19 | https://s.a80.fr/conversation/44                      |   1 |
|  11 | https://qvitter.krashboyz.org/conversation/41767      |   1 |
| 417 | https://qvitter.krashboyz.org/conversation/41767      |   1 |
| 262 | https://ostatus.shnoulle.net/conversation/577334      |   1 |
|   9 | https://qvitter.krashboyz.org/conversation/41767      |   1 |
| 416 | https://qvitter.krashboyz.org/conversation/41767      |   1 |
|   7 | https://qvitter.krashboyz.org/conversation/41767      |   1 |
| 415 | https://qvitter.krashboyz.org/conversation/41767      |   1 |
| 260 | https://ostatus.shnoulle.net/conversation/577334      |   1 |
|   5 | https://qvitter.krashboyz.org/conversation/41767      |   1 |
| 414 | https://qvitter.krashboyz.org/conversation/41767      |   1 |
| 258 | https://ostatus.shnoulle.net/conversation/577334      |   1 |
| 256 | https://ostatus.shnoulle.net/conversation/577334      |   1 |
| 254 | https://ostatus.shnoulle.net/conversation/577334      |   1 |
| 226 | https://social.umeahackerspace.se/conversation/660093 |   1 |
| 223 | https://social.umeahackerspace.se/conversation/660093 |   1 |
| 150 | https://social.umeahackerspace.se/conversation/659945 |   1 |
| 140 | https://social.umeahackerspace.se/conversation/659617 |   1 |
| 138 | https://social.umeahackerspace.se/conversation/659617 |   1 |
| 136 | https://social.umeahackerspace.se/conversation/659617 |   1 |
| 148 | https://social.umeahackerspace.se/conversation/659945 |   1 |
| 134 | https://social.umeahackerspace.se/conversation/659617 |   1 |
| 132 | https://social.umeahackerspace.se/conversation/659617 |   1 |
| 144 | https://social.umeahackerspace.se/conversation/659981 |   1 |
| 130 | https://social.umeahackerspace.se/conversation/659617 |   1 |
| 128 | https://social.umeahackerspace.se/conversation/659617 |   1 |
| 425 | https://social.umeahackerspace.se/conversation/659945 |   1 |
| 427 | https://social.umeahackerspace.se/conversation/659981 |   1 |
| 126 | https://social.umeahackerspace.se/conversation/659617 |   1 |
| 142 | https://social.umeahackerspace.se/conversation/659981 |   1 |
| 212 | https://herds.eu/conversation/1921352                 |   1 |
| 210 | https://herds.eu/conversation/1921352                 |   1 |
| 124 | https://social.umeahackerspace.se/conversation/659617 |   1 |
| 146 | https://social.umeahackerspace.se/conversation/659945 |   1 |
| 408 | https://social.umeahackerspace.se/conversation/659617 |   1 |
| 208 | https://herds.eu/conversation/1921352                 |   1 |
+-----+-------------------------------------------------------+-----+
162 rows in set (0,01 sec)
````

Alkarex commented:
One more:

```
mysql> SELECT count(`event`.`id`) AS total, type, start, adjust FROM `event` WHERE `event`.`uid` = 1 AND `start` < '2017-04-19 12:00:06' AND `finish` > '2017-04-12 12:00:06' and `ignore` = 0 ORDER BY `start` ASC;
ERROR 1140 (42000): In aggregated query without GROUP BY, expression #2 of SELECT list contains nonaggregated column 'friendica.event.type'; this is incompatible with sql_mode=only_full_group_by
```

Alkarex commented:
I have addressed a bunch of the easy GROUP BY fixes in https://github.com/friendica/friendica/pull/3323/commits/74b6d09e89bbe87e94ebb458d8a76415e324d1d3 (the queries which I had a possibility to test manually, excluding the ones which are too much dynamically generated)

Alkarex commented:
Most changes should be safe. https://github.com/friendica/friendica/pull/3323/commits/c3e933642e6ad761c91888006b45efa69915d691 would benefit from a little extra review.

Hypolite commented on mod/ping.php:
```diff
@@ -201,6 +201,7 @@ function ping_init(App $a)
 		if (is_null($ev)) {
 			$ev = qu("SELECT count(`event`.`id`) AS total, type, start, adjust FROM `event`
 				WHERE `event`.`uid` = %d AND `start` < '%s' AND `finish` > '%s' and `ignore` = 0
+				GROUP BY type, start, adjust
```
As suggested by line 219-220 below, `$ev` is supposed to be an array of the events in the next 7 days. The `COUNT` seems to be superfluous here, as it can be obtained by doing `count($ev)`.

Alkarex commented:
Another request fixed (profile wrong DISTINCT + ORDER BY) in https://github.com/friendica/friendica/pull/3323/commits/9fb4ba2b733d59b8a704cb8df846c3b43769503d

    ERROR 3065 (HY000) Expression #1 of ORDER BY clause is not in
    SELECT list, references column 'friendica.item.created' which is not in
    SELECT list; this is incompatible with DISTINCT

annando commented:
@Alkarex we desperately need someone who could have a look at our SQL queries. It will be a very hard job, compare it with the cleaning of a toilet at a train station ;-) You will find some very obscure queries.

It would be really great if you could have a look.

A friend created a nice shell script for me. Her script lists all queries:
```
#!/bin/bash
LIST=./list.txt

for file in `find -type f -regex ".*\.php$"`; do \

        grep -e 'q(' -e 'qu(' "$file" 1>/dev/null

        if [[ "$?" == "0" ]]; then

                echo -e "-------------------\n$file\n-------------------\n\n" >> $LIST
                sed -n '/\(q\|qu\)(/','/);/{=;p}' $file >> $LIST
                echo -e "\n" >> $LIST
        fi
done

sed -i '/^[0-9]\+$/{N;s/\n//;}' $LIST
```
Although I'm the one who is doing the whole database stuff, I'm only a rookie with some experience.

Alkarex commented:
:-) I am fixing the SQL requests that fail in standard MySQL (5.7.17). The requests cannot be taken out of the PHP code, because the complex ones are dynamically generated (which is understandable for such as system). For instance, I am having a look at this GROUP BY, which is wrong, but hard to fix because the terms SELECT are dynamic: https://github.com/friendica/friendica/blob/7b352f3f74b045690cdeae507b609f7a9cc8db03/mod/search.php#L213-L220

Alkarex commented:
The issue from my post above should be fixed in https://github.com/friendica/friendica/pull/3323/commits/38e7a0f7933461702275a846a85e9dc3ae5dd97e

Alkarex commented:
While adding GROUP BY on a random field might change the nature of the returned results, the changes I have made are not random (I hope). In particular, adding a GROUP BY on a field that is a unique key is always safe https://github.com/friendica/friendica/pull/3323/commits/38e7a0f7933461702275a846a85e9dc3ae5dd97e 

A few reasons not to use ANY_VALUE:
* It is specific to MySQL (not available in e.g. SQLite, PostgreSQL, which might be considered for Friendica at some point)
* It is not available in MySQL 5.6- https://dev.mysql.com/doc/refman/5.6/en/miscellaneous-functions.html
* It pushes the problem of clearly stating the wanted SQL request, using a non-deterministic fix

Alkarex commented:
So my suggestion is to test my patch on a old version of MySQL and check that the behaviour is as expected. In any case, I think those SQL requests must be fixed, as they have never been standard SQL.
You can see a few examples of outputs higher up https://github.com/friendica/friendica/pull/3323#issuecomment-293526197

Alkarex commented:
If you want, what I could do is to go through the changes, and group by a unique key more often than the changes that I have done so far.

Alkarex commented:
Ok, you are right, and I have been a bit over-optimistic :-) I think I now get what the old behaviour was for non-standard requests. But still, what I am advocating is to go back and stick to standard SQL. The current version of several requests is not standard. Fixing / standardising them will make them work in all versions of MySQL (as well as other databases).

The difficulty (for me) is to find out what the intention was with the original non-standard SQL queries... For instance, with your example, what is the intention? If it is to get distinct URLs, then one can state it (e.g. with DISTINCT). Is there a reason to want random values in the SELECT?

A few tests:

```sql
-- The first one is not standard SQL
SELECT * FROM `item`
GROUP BY `item`.`uri`;
ERROR 1055 (42000): Expression #1 of SELECT list is not in GROUP BY clause and contains nonaggregated column 'friendica.item.id' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by

SELECT uri FROM `item`;
1137 rows in set (0,01 sec)

SELECT uri FROM `item` group by uri;
593 rows in set (0,00 sec)

SELECT uri FROM `item` group by uri, id;
1137 rows in set (0,01 sec)

SELECT uri FROM `item` group by uri, id, guid, `contact-id`;
1137 rows in set (0,00 sec)
-- Adding more fields in a group by after a unique key should not change the response

SELECT DISTINCT uri FROM `item`;
593 rows in set (0,00 sec)
```

Hypolite commented:
I am 100% on board with standard SQL, however it shouldn't change the current behavior of queries.

About your question about random values, in the `item` table there is a lot a duplicated content between public items and items delivered to specific accounts. So in most cases we don't care which row the value comes from because they probably are identical or close enough that it isn't relevant, like the sorting on `item.id` after grouping on `item.uri`.

Alkarex commented:
Would you be able to find out which of the modified queries actually expect this ANY_VALUE behaviour you just described? The other, deterministic queries are easier to fix.

Hypolite commented:
I will conduct a more thorough review of the affected queries this evening EST.

Alkarex commented:
For the record, it does not look like MariaDB has ANY_VALUE() yet https://jira.mariadb.org/browse/MDEV-10426

Hypolite commented:
What about the ONLY_FULL_GROUP default setting?

Alkarex commented:
Just brainstorming: for the queries that actually expect the behaviour of ANY_VALUE(), a quick fix could be to use an aggregate function such as MIN(), which should work everywhere. If there is a desire to optimise at the cost of standardisation, the aggregate function could be picked by PHP between "ANY_VALUE()", "MIN()", and "()" (nothing) depending on the database. E.g.:

```sql
SELECT uri, min(id), min(guid), min(`contact-id`) FROM `item` group by uri;
598 rows in set (0,00 sec)
```

annando commented:
Standard SQL sounds good - when the result is identical.

Hypolite commented:
I believe this could work, especially in the `ORDER BY` case.

Hypolite commented on mod/search.php:
```diff
@@ -214,7 +214,7 @@ function search_content(App $a) {
 			WHERE %s AND (`item`.`uid` = 0 OR (`item`.`uid` = %s AND NOT `item`.`global`))
 				$sql_extra
-			GROUP BY `item`.`uri` ORDER BY `item`.`id` DESC LIMIT %d , %d",
+			GROUP BY `item`.`uri`, `item`.`id` ORDER BY `item`.`id` DESC LIMIT %d , %d",
```
✖️ I suggest using `MIN(item.id) AS item_id` in the `SELECT` and `ORDER BY item_id`.

Alkarex replied:
If we do not GROUP BY a unique key such as item.id, then we must use MIN() on all the SELECT fields. Furthermore, item.id is already selected twice in https://github.com/friendica/friendica/blob/7b352f3f74b045690cdeae507b609f7a9cc8db03/include/conversation.php#L421-L430 :-)

Hypolite replied:
My suggestion was very indeed generic and at the end of an exhausting review process, so it definitely can't apply as is to this query. However there's no way around it, this group by isn't working either.

I have no good easy simple solution to offer you, the query clearly wasn't designed with respect to SQL92, and we are left with the ungrateful job of fixing stuff.

Alkarex replied:
@Hypolite Could you please have a look again at this one? The additional GROUP BY seems to give fine results on my instance.

Hypolite replied:
Confirmed, was it the last pending change?

Alkarex replied:
The last one of the requests which were crashing on my system, yes. I have not reviewed all requests.

Hypolite replied:
Ok, would you prefer we merge this PR now or do you want to do an exhaustive query review before?

Alkarex replied:
In my opinion, it would be good if one of you could test this pull request on your system and then merge. Going through all requests will likely be a slightly longer process.

Hypolite commented on mod/videos.php:
```diff
@@ -358,7 +358,7 @@ function videos_content(App $a) {
 	$r = q("SELECT hash, `id`, `filename`, filetype FROM `attach`
 		WHERE `uid` = %d AND filetype LIKE '%%video%%'
-		$sql_extra GROUP BY hash ORDER BY `created` DESC LIMIT %d , %d",
+		$sql_extra GROUP BY hash, `id` ORDER BY `created` DESC LIMIT %d , %d",
```
✖️ I suggest adding `MIN(id) AS id, MIN(created) AS created` in the `SELECT`

Alkarex replied:
For reference, EXPLAIN output (the version with ANY_VALUE should be the current behaviour before the change):
```
mysql> EXPLAIN SELECT hash, MIN(`id`) AS `id`, MIN(`created`) AS `created`, min(`filename`) AS `filename`, min(`filetype`) as `filetype` FROM `attach`
    -> GROUP BY hash ORDER BY `created` DESC;
+----+-------------+--------+------------+------+---------------+------+---------+------+------+----------+---------------------------------+
| id | select_type | table  | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra                           |
+----+-------------+--------+------------+------+---------------+------+---------+------+------+----------+---------------------------------+
|  1 | SIMPLE      | attach | NULL       | ALL  | NULL          | NULL | NULL    | NULL |    1 |   100.00 | Using temporary; Using filesort |
+----+-------------+--------+------------+------+---------------+------+---------+------+------+----------+---------------------------------+
1 row in set, 1 warning (0,06 sec)

mysql> EXPLAIN SELECT hash, ANY_VALUE(`id`) AS `id`, ANY_VALUE(`created`) AS `created`, ANY_VALUE(`filename`) AS `filename`, ANY_VALUE(`filetype`) as `filetype` FROM `attach`
    -> GROUP BY hash ORDER BY `created` DESC;
+----+-------------+--------+------------+------+---------------+------+---------+------+------+----------+---------------------------------+
| id | select_type | table  | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra                           |
+----+-------------+--------+------------+------+---------------+------+---------+------+------+----------+---------------------------------+
|  1 | SIMPLE      | attach | NULL       | ALL  | NULL          | NULL | NULL    | NULL |    1 |   100.00 | Using temporary; Using filesort |
+----+-------------+--------+------------+------+---------------+------+---------+------+------+----------+---------------------------------+
1 row in set, 1 warning (0,00 sec)
```

annando commented:
By everything you do, please also have a look what ```explain```is telling about the changed queries. I wouldn't like to have some new queries that now are doing a full table scan (and they didn't before).

Alkarex commented on include/api.php:
```diff
@@ -3102,8 +3102,10 @@ function api_fr_photo_detail($type) {
+					`type`, `height`, `width`, `datasize`, `profile`, min(`scale`) as minscale, max(`scale`) as maxscale
+				from photo where `uid` = %d and `resource-id` = '%s' %s
+				group by `resource-id`, `created`, `edited`, `title`, `desc`, `album`, `filename`,
+					`type`, `height`, `width`, `datasize`, `profile`",
```
```
explain select ANY_VALUE(`resource-id`), ANY_VALUE(`created`), ANY_VALUE(`edited`), ANY_VALUE(`title`), min(`scale`) as minscale, max(`scale`) as maxscale
    -> from photo where `uid` = 1 and `resource-id` = '3daafed7834e357d2af2b0bcdb8fe468';
+----+-------------+-------+------------+------+---------------------------------------------------------------------------------------------+-------------+---------+-------+------+----------+-------------+
| id | select_type | table | partitions | type | possible_keys                                                                               | key         | key_len | ref   | rows | filtered | Extra       |
+----+-------------+-------+------------+------+---------------------------------------------------------------------------------------------+-------------+---------+-------+------+----------+-------------+
|  1 | SIMPLE      | photo | NULL       | ref  | uid_contactid,uid_profile,uid_album_scale_created,uid_album_resource-id_created,resource-id | resource-id | 258     | const |    3 |     7.67 | Using where |
+----+-------------+-------+------------+------+---------------------------------------------------------------------------------------------+-------------+---------+-------+------+----------+-------------+
1 row in set, 1 warning (0,03 sec)

mysql> explain select MIN(`resource-id`), MIN(`created`), MIN(`edited`), MIN(`title`), min(`scale`) as minscale, max(`scale`) as maxscale from photo where `uid` = 1 and `resource-id` = '3daafed7834e357d2af2b0bcdb8fe468';
+----+-------------+-------+------------+------+---------------------------------------------------------------------------------------------+-------------+---------+-------+------+----------+-------------+
| id | select_type | table | partitions | type | possible_keys                                                                               | key         | key_len | ref   | rows | filtered | Extra       |
+----+-------------+-------+------------+------+---------------------------------------------------------------------------------------------+-------------+---------+-------+------+----------+-------------+
|  1 | SIMPLE      | photo | NULL       | ref  | uid_contactid,uid_profile,uid_album_scale_created,uid_album_resource-id_created,resource-id | resource-id | 258     | const |    3 |     7.67 | Using where |
+----+-------------+-------+------------+------+---------------------------------------------------------------------------------------------+-------------+---------+-------+------+----------+-------------+
1 row in set, 1 warning (0,00 sec)
```

Alkarex commented:
Except for mod/search.php (waiting to find a good approach), I have rewritten the queries to use `ANY_VALUE()` as this is the semantic operator for what is actually expected currently for those requests.
I have introduced a new function to change `ANY_VALUE()` to `MIN()` which has the advantage of working everywhere https://github.com/Alkarex/friendica/blob/6be6badd915800582ddeaf5b6f9162623205e868/include/dba.php#L507-L513
This function could be optimised if we know on which database the system is running.
Let me know what you think, and it would be great if you could give it a try.

Hypolite commented on include/dba.php:
```diff
@@ -504,6 +504,14 @@ function dbesc($str) {
+	//ANY_VALUE() is available from MySQL 5.7 https://dev.mysql.com/doc/refman/5.7/en/miscellaneous-functions.html
+	//A standard fallback is to use MIN(), or nothing () in old MySQL 5.6-
+	//TODO: Skip this line when we know we are on a platform supporting ANY_VALUE()
+	return str_ireplace('ANY_VALUE(', 'MIN(', $sql);
```
You can use something similar with what we use in [`include/dbstructure.php`](https://github.com/friendica/friendica/blob/develop/include/dbstructure.php#L184) to check for a specific version of MySQL.

annando replied:
I wouldn't like to have new functions in dba.php. Instead I would like to have them all in dbm.php

annando replied:
Additionally I'm not happy with this function. Doesn't it simply replaces the one text with the other one? So when we are discussing about this topic in Friendica then even inside of the body message the "any_value" will be replaced with "min".

Alkarex replied:
@annando The string replace is supposed to be done in the SQL query excluding its values/parameters.

Alkarex replied:
@annando I can move this new function elsewhere, but as it is only used by `q()` and `qu()`, I thought it made sense to have it next to them.

Alkarex replied:
@Hypolite Version detection implemented in https://github.com/friendica/friendica/pull/3323/commits/5007be3cf55d5f4854a11b06811da9e50c256829

annando replied:
When it is only used by "q" you could also make it a function of the database class. But I wouldn't like it to be some standalone function out there in the wild.

Alkarex replied:
@annando Done in https://github.com/friendica/friendica/pull/3323/commits/062070058e48c584616493af38d9949eccd9a849

rabuzarus commented on include/dba.php:
```diff
@@ -504,6 +504,14 @@ function dbesc($str) {
 	}
 }
 
+function any_value_fallback($sql) {
```
STANDARDS: Please add a doxygen header

Alkarex replied:
Done in https://github.com/friendica/friendica/pull/3323/commits/06f374b26b01674707c135843d0df5189a9e57cf

Alkarex commented:
On more small SQL request patched in https://github.com/friendica/friendica/pull/3323/commits/8b3f623d29e7228115441d4824e2893a1585fe15
The query was not deterministic, and therefore was not doing what its description says. The patch makes it a bit more standard and deterministic, but the returned protocol type is still not deterministic (I am not sure it is a problem if it has worked until now).

annando commented:
There are problems with this pull request: https://helpers.pyxis.uberspace.de/display/0b6b25a83458f3d8c1601a9508484005

Alkarex commented:
@annando Do you know which request(s) are involved in that bug?

Hypolite commented:
@Alkarex I expect it to be the query in `include/notifier.php`, but I'm not sure.

rabuzarus commented:
No, it is the private messages page. So I would say you need to check the queries in `mod/message.php`

Alkarex commented:
I have found the problem. I had forgotten the `AS` in that request. Patch coming.

Alkarex commented:
Should be fixed in https://github.com/friendica/friendica/pull/3348
